### ISSUE TITLE
retry from default options for queries is overwritten with undefined retry in useQuery

### ISSUE THREAD

**Issue by vdvukhzhilov** [TanStack/query#7472]
### Describe the bug

You can provide `retry` in `defaultOptions.queries` when instantiating `QueryClient`.
This is used when no `retry` option is provided to `useQuery`.

But when we provide `retry: undefined` to `useQuery` options `defaultOptions.queries.retry` is ignored, instead lib default retry kicks in.

### Your minimal, reproducible example

https://codesandbox.io/p/sandbox/react-query-retry-bug-vd8y9h?file=%2Fsrc%2FApp.js%3A5%2C31

### Steps to reproduce

1. Add any logic or retry count to default options of a query client
2. Add `retry: undefined` as option to useQuery

### Expected behavior

`retry` is used from `defaultOptions.queries`

### How often does this bug happen?

Every time

### Platform

any platfrom

### Tanstack Query adapter

None

### TanStack Query version

v5.37.1

**Comment by vdvukhzhilov** [comment]
I believe the problem is in [this line](https://github.com/TanStack/query/blob/main/packages/query-core/src/query.ts#L188)

`undefined` values should be filtered out before assigning as they are essentially the same as providing no value at all and in that case should fallback to defaultOptions and than fallback to lib defaults

**Comment by vdvukhzhilov** [comment]
More advanced example

```
function useSomeQuery({ retry }: {retry?: number}) {
  return useQuery({
    retry
  })
}
```

Here we suppose that you can omit retry when passing to `useSomeQuery({})`. But due to spread of argument retry is passed as `undefined` to `useQuery`

**Comment by vdvukhzhilov** [comment]
> The default options only kick in when the key is not passed to useQuery...

Well, If that is intended, it should be pointed out in documentation. Otherwise, it is totally unclear because of semantics of term "default options".

**Comment by TkDodo** [comment]
> If that is by design, I would say that this should be pointed out in documentation.

sure, feel free to amend the docs.

> have not found any information in docs or your blog that would explain this particular behaviour 

it's not like this question has come up a lot - this is the first time in 4 years for me 😅 . So I guess most people just don't pass `something: undefined` to `useQuery`.

> I have already shown how it could be view as "not providing" option

Do you mean this example?

```
function useSomeQuery({ retry }: {retry?: number}) {
  return useQuery({
    retry
  })
}
```

Because with [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes) turned on, you couldn't pass `retry:undefined` to your custom hook. `{ retry?: number }` is different from `{ retry: number | undefined }`, and since you clearly care about this distinction, I'd suggest turning that TypeScript feature on.

**Comment by vdvukhzhilov** [comment]
> [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes)

I agree that would help to prevent explicitly passing `undefined` as value, but it not the case here.
The ones using this custom hook do not pass undefined like `useSomeQuery({ retry: undefined })`, they do `useSomeQuery({})`. It is the way speading objects works in JS that results in `retry` variable inside `useSomeQuery` being defined (with value `undefined`).

Basically
```
function useSomeQuery({ retry }: {retry?: number}) {
  return useQuery({
    retry
  })
}
```
And
```
function useSomeQuery(props: {retry?: number}) {
  const retry = props.retry;
  return useQuery({
    retry
  })
}
```
is the same.
